File: repl.ts
import * as console from "node:console";
import { stdin, stdout } from "node:process";
import { createInterface } from "node:readline";
import { evaluate } from "./evaluator.ts";
import { Lexer } from "./lexer.ts";
import { Parser } from "./parser.ts";

const repl = createInterface({
  input: stdin,
  output: stdout,
  prompt: ">> ",
});

repl.on("line", (line) => {
  const lexer = new Lexer(line);
  const parser = new Parser(lexer);

  const program = parser.parseProgram();
  if (parser.errors.length !== 0) {
    for (let i = 0; i < parser.errors.length; i++) {
      console.error(parser.errors[i]);
    }
  } else {
    const evaluated = evaluate(program);
    if (evaluated !== null) {
      console.log(evaluated.inspect());
    }
  }

  repl.prompt();
});

repl.on("close", () => {
  repl.close();
});

repl.prompt();

File: go/repl/repl.go
package repl

import (
	"bufio"
	"fmt"
	"io"

	"github.com/brattonross/monkeylang/evaluator"
	"github.com/brattonross/monkeylang/lexer"
	"github.com/brattonross/monkeylang/object"
	"github.com/brattonross/monkeylang/parser"
)

const PROMPT = ">> "

func Start(r io.Reader, w io.Writer) {
	scanner := bufio.NewScanner(r)
	env := object.NewEnvironment()

	for {
		fmt.Fprint(w, PROMPT)
		scanned := scanner.Scan()
		if !scanned {
			return
		}

		line := scanner.Text()
		l := lexer.New(line)
		p := parser.New(l)

		program := p.ParseProgram()
		if len(p.Errors()) > 0 {
			printParserErrors(w, p.Errors())
			continue
		}

		evaluated := evaluator.Eval(program, env)
		if evaluated != nil {
			io.WriteString(w, evaluated.Inspect())
			io.WriteString(w, "\n")
		}
	}
}

func printParserErrors(w io.Writer, errors []string) {
	fmt.Fprintf(w, "Woops! We ran into some monkey business here!\n")
	fmt.Fprintf(w, " parser errors:\n")
	for _, msg := range errors {
		fmt.Fprintf(w, "\t%s\n", msg)
	}
}
